packages/core/src/plugins/DependencyManager.ts

```diff
@@ -0,0 +1,52 @@
+import type { Constructor } from '../utils/mixins'
+import type { ModulePlugin, Plugin } from './Plugin'
+import type { DependencyContainer } from 'tsyringe'
+
+import { InjectionToken, Lifecycle } from 'tsyringe'
+
+export { InjectionToken }
+
+export class DependencyManager {
+  public container: DependencyContainer
+
+  public constructor(container: DependencyContainer) {
+    this.container = container
+  }
+
+  public registerModulePlugins(modules: ModulePlugin | ModulePlugin[]) {
+    const modulesArray = Array.isArray(modules) ? modules : [modules]
+
+    modulesArray.forEach((module) => {
+      // Register Module class to be instantiated per container
+      this.container.register(module, module, { lifecycle: Lifecycle.ContainerScoped })
+
+      // Register all dependencies of this module
+      module.register(this)
+    })
+  }
+
+  public registerPlugins(plugins: Plugin | Plugin[]) {
+    const pluginsArray = Array.isArray(plugins) ? plugins : [plugins]
+
+    pluginsArray.forEach((plugin) => plugin.register(this))
+  }
+
+  public registerSingleton<T>(from: InjectionToken<T>, to: InjectionToken<T>): void
+  public registerSingleton<T>(token: Constructor<T>): void
```

Comment:
Is this function overloading? Looks cool!

Reply:
Yes! TSyringe uses it so to match their API we have to add these overloading calls